<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Deconvolution and decomposition of mouse cortex with Spotiphy &mdash; Spotiphy 0.1.2 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=92fd9be5" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=19f00094" />
      <link rel="stylesheet" type="text/css" href="_static/nbsphinx-code-cells.css" />

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js?v=5d32c60e"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="_static/documentation_options.js?v=92734c54"></script>
        <script src="_static/doctools.js?v=888ff710"></script>
        <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Frequently Asked Questions" href="questions.html" />
    <link rel="prev" title="Tutorials" href="tutorials.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            Spotiphy
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Basic</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="tutorials.html">Tutorials</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">Deconvolution and decomposition of mouse cortex with Spotiphy</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Contents">Contents</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Install-and-load-packages">Install and load packages</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Load-Visium-data,-scRNA-data,-and-the-HE-staining-image.">Load Visium data, scRNA data, and the HE staining image.</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Select-marker-genes-for-the-deconvolution">Select marker genes for the deconvolution</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Deconvolution:-Estimate-cell-abundance-in-spatial-spots">Deconvolution: Estimate cell abundance in spatial spots</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Segmentation:-Estimate-cell-number-in-each-spatial-spot">Segmentation: Estimate cell number in each spatial spot</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Decomposition:-Decompose-the-expression-of-each-spatial-spot-into-single-cell-level">Decomposition: Decompose the expression of each spatial spot into single-cell level</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Generate-pseudo-single-cell-resolution-image">Generate pseudo single-cell resolution image</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="questions.html">Frequently Asked Questions</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="scRNA%20Reference.html">scRNA Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="Deconvolution%20and%20Decomposition.html">Deconvolution and Decomposition</a></li>
<li class="toctree-l1"><a class="reference internal" href="Segmentation.html">Segmentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="Visualization.html">Visualization</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">About</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="About.html">About the project</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Spotiphy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="tutorials.html">Tutorials</a></li>
      <li class="breadcrumb-item active">Deconvolution and decomposition of mouse cortex with Spotiphy</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/Spotiphy_tutorial_1.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="Deconvolution-and-decomposition-of-mouse-cortex-with-Spotiphy">
<h1>Deconvolution and decomposition of mouse cortex with Spotiphy<a class="headerlink" href="#Deconvolution-and-decomposition-of-mouse-cortex-with-Spotiphy" title="Link to this heading">ÔÉÅ</a></h1>
<ul class="simple">
<li><p>by Ziqian Zheng <a class="reference external" href="mailto:zzheng92&#37;&#52;&#48;wisc&#46;edu">zzheng92<span>&#64;</span>wisc<span>&#46;</span>edu</a> and Jiyuna Yang <a class="reference external" href="mailto:jiyuan&#46;yang&#37;&#52;&#48;stjude&#46;org">jiyuan<span>&#46;</span>yang<span>&#64;</span>stjude<span>&#46;</span>org</a>.</p></li>
<li><p>Last update: Nov 29th 2023</p></li>
</ul>
<section id="Contents">
<h2>Contents<a class="headerlink" href="#Contents" title="Link to this heading">ÔÉÅ</a></h2>
<ol class="arabic simple">
<li><p>Install and load packages</p></li>
<li><p>Load Visium and scRNA data</p></li>
<li><p>Select marker genes for the deconvolution</p></li>
<li><p>Deconvolution: Estimate cell abundance in spatial spots</p></li>
<li><p>Segmentation: Estimate cell number in each spatial spot</p></li>
<li><p>Decomposition: Decompose the expression of each spatial spot into single-cell level</p></li>
<li><p>Generate pseudo single-cell resolution image</p></li>
</ol>
</section>
<section id="Install-and-load-packages">
<h2>Install and load packages<a class="headerlink" href="#Install-and-load-packages" title="Link to this heading">ÔÉÅ</a></h2>
<p>If you‚Äôre using you local environment, please ensure that <code class="docutils literal notranslate"><span class="pre">spotiphy</span></code> is correctly installed by following the instructions on its <a class="reference external" href="https://github.com/jyyulab/Spotiphy">GitHub page</a>.</p>
<p>If you‚Äôre using Google Colab, please execute the following code block to install <code class="docutils literal notranslate"><span class="pre">spotiphy</span></code> and all its dependencies. To ensure the code runs smoothly in Google Colab, we will manually eliminate some unnecessary Jupyter variables.</p>
<p><strong>üòÆImportant:</strong> To get enough RAM, <a class="reference external" href="https://colab.research.google.com/signup">Colab Pro</a> might be needed.</p>
<p>For the first time the following code block is executed in Google Colab, you may get an error of inconsistent package versions like this:</p>
<ul class="simple">
<li><p>ERROR: pip‚Äôs dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.</p></li>
</ul>
<p>To solve this issue, simply ignore this error and <strong>restart the session</strong> (Click <em>Runtime</em>, then click <em>Restart Session</em>). Then you will be able to load the packages correctly.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">sys</span>
<span class="n">IN_COLAB</span> <span class="o">=</span> <span class="s2">&quot;google.colab&quot;</span> <span class="ow">in</span> <span class="n">sys</span><span class="o">.</span><span class="n">modules</span>
<span class="k">if</span> <span class="n">IN_COLAB</span><span class="p">:</span>
    <span class="err">!</span><span class="n">pip</span> <span class="n">install</span> <span class="o">-</span><span class="n">qq</span> <span class="n">spotiphy</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">spotiphy</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">pickle</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">matplotlib</span> <span class="k">as</span> <span class="nn">mpl</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">torch</span>
<span class="kn">import</span> <span class="nn">cv2</span>

<span class="n">results_folder</span> <span class="o">=</span> <span class="s1">&#39;results/mouse_brain/&#39;</span>
<span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">results_folder</span><span class="p">):</span>
  <span class="c1"># Create result folder if it does not exist</span>
  <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">results_folder</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="Load-Visium-data,-scRNA-data,-and-the-HE-staining-image.">
<h2>Load Visium data, scRNA data, and the HE staining image.<a class="headerlink" href="#Load-Visium-data,-scRNA-data,-and-the-HE-staining-image." title="Link to this heading">ÔÉÅ</a></h2>
<p>Verify the existence of the required data. If they are not present, download them from GitHub.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">requests</span>
<span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="s1">&#39;data/&#39;</span><span class="p">):</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="s1">&#39;data/&#39;</span><span class="p">)</span>

<span class="n">data_path</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;data/scRNA_mouse_brain.h5ad&#39;</span><span class="p">,</span> <span class="s1">&#39;data/ST_mouse_brain.h5ad&#39;</span><span class="p">,</span>
             <span class="s1">&#39;data/img_mouse_brain.png&#39;</span><span class="p">]</span>
<span class="n">urls</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;https://raw.githubusercontent.com/jyyulab/Spotiphy/main/tutorials/data/scRNA_mouse_brain.h5ad&#39;</span><span class="p">,</span>
        <span class="s1">&#39;https://raw.githubusercontent.com/jyyulab/Spotiphy/main/tutorials/data/ST_mouse_brain.h5ad&#39;</span><span class="p">,</span>
        <span class="s1">&#39;https://raw.githubusercontent.com/jyyulab/Spotiphy/main/tutorials/data/img_mouse_brain.png&#39;</span><span class="p">]</span>

<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">path</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">data_path</span><span class="p">):</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
        <span class="n">response</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">urls</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">allow_redirects</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="s1">&#39;wb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
            <span class="n">file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">response</span><span class="o">.</span><span class="n">content</span><span class="p">)</span>
        <span class="k">del</span> <span class="n">response</span>
</pre></div>
</div>
</div>
<p>Load the scRNA data, visium data and the corresponding H&amp;E staining image.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">adata_sc</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s2">&quot;data/scRNA_mouse_brain.h5ad&quot;</span><span class="p">)</span>
<span class="n">adata_st</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s2">&quot;data/ST_mouse_brain.h5ad&quot;</span><span class="p">)</span>
<span class="n">img</span> <span class="o">=</span> <span class="n">cv2</span><span class="o">.</span><span class="n">imread</span><span class="p">(</span><span class="s2">&quot;data/img_mouse_brain.png&quot;</span><span class="p">)</span>

<span class="n">key_type</span> <span class="o">=</span> <span class="s1">&#39;celltype&#39;</span>
<span class="n">type_list</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">adata_sc</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">key_type</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()))</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;There are </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">type_list</span><span class="p">)</span><span class="si">}</span><span class="s1"> cell types: </span><span class="si">{</span><span class="n">type_list</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
There are 14 cell types: [&#39;Astro&#39;, &#39;CA&#39;, &#39;DG&#39;, &#39;L2/3 IT CTX&#39;, &#39;L4 IT CTX&#39;, &#39;L4/5 IT CTX&#39;, &#39;L5 IT CTX&#39;, &#39;L5 PT CTX&#39;, &#39;L5/6 IT CTX&#39;, &#39;L6 CT CTX&#39;, &#39;L6b CTX&#39;, &#39;Microglia&#39;, &#39;Oligo&#39;, &#39;SUB&#39;]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">mpl</span><span class="o">.</span><span class="n">pyplot</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spatial</span><span class="p">(</span><span class="n">adata_st</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;in_tissue&quot;</span><span class="p">,</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata_sc</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;celltype&quot;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">img</span><span class="p">[:,</span> <span class="p">:,</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">]])</span>  <span class="c1"># In default, cv2 uses BGR. So we convert BGR to RGB.</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">mpl</span><span class="o">.</span><span class="n">pyplot</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
scatterplots.py (392): No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/Spotiphy_tutorial_1_9_1.png" src="_images/Spotiphy_tutorial_1_9_1.png" />
</div>
</div>
</section>
<section id="Select-marker-genes-for-the-deconvolution">
<h2>Select marker genes for the deconvolution<a class="headerlink" href="#Select-marker-genes-for-the-deconvolution" title="Link to this heading">ÔÉÅ</a></h2>
<p>Initialize the spatial trancriptomics and scRNA data. Specifically, - <code class="docutils literal notranslate"><span class="pre">adata_sc</span></code> and <code class="docutils literal notranslate"><span class="pre">adata_st</span></code> are normalized - cells and genes in <code class="docutils literal notranslate"><span class="pre">adata_sc</span></code> are filtered - only the common genes shared by <code class="docutils literal notranslate"><span class="pre">adata_sc</span></code> and <code class="docutils literal notranslate"><span class="pre">adata_st</span></code> are kept</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">adata_sc</span><span class="p">,</span> <span class="n">adata_st</span> <span class="o">=</span> <span class="n">spotiphy</span><span class="o">.</span><span class="n">initialization</span><span class="p">(</span><span class="n">adata_sc</span><span class="p">,</span> <span class="n">adata_st</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Convert expression matrix to array: 0.98s
Normalization: 3.52s
Filtering: 3.8s
Find common genes: 0.05s
</pre></div></div>
</div>
<p>Select marker genes based on pairwise statistical tests. The selected marker genes are saved as a CSV file.</p>
<p>Users can adjust the following arguments:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">n_select</span></code>: the maximum number of marker genes select for each cell type</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">threshold_p</span></code>: threshold of the <span class="math notranslate nohighlight">\(p\)</span>-value to be considered significant</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">threshold_fold</span></code>: threshold of the fold-change to be considered significant</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">q</span></code>: Since more than one <span class="math notranslate nohighlight">\(p\)</span>-values are obtained in using the pairwise tests, we only consider the <span class="math notranslate nohighlight">\(p\)</span>-value at quantile <code class="docutils literal notranslate"><span class="pre">q</span></code>. For additional details, please refer to the supplementary material of the paper.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">return_dict</span></code>: Whether return a dictionary representing the marker genes selected for each cell type, or just return the list of all selected marker genes.</p></li>
</ul>
<p>The parameters specified in the following code block provide a good starting point.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">marker_gene_dict</span> <span class="o">=</span> <span class="n">spotiphy</span><span class="o">.</span><span class="n">sc_reference</span><span class="o">.</span><span class="n">marker_selection</span><span class="p">(</span><span class="n">adata_sc</span><span class="p">,</span> <span class="n">key_type</span><span class="o">=</span><span class="n">key_type</span><span class="p">,</span> <span class="n">return_dict</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                                                          <span class="n">n_select</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">threshold_p</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">threshold_fold</span><span class="o">=</span><span class="mf">1.5</span><span class="p">,</span>
                                                          <span class="n">q</span><span class="o">=</span><span class="mf">0.15</span><span class="p">)</span>
<span class="n">marker_gene</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">marker_gene_label</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">type_</span> <span class="ow">in</span> <span class="n">type_list</span><span class="p">:</span>
    <span class="n">marker_gene</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">marker_gene_dict</span><span class="p">[</span><span class="n">type_</span><span class="p">])</span>
    <span class="n">marker_gene_label</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="n">type_</span><span class="p">]</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">marker_gene_dict</span><span class="p">[</span><span class="n">type_</span><span class="p">]))</span>
<span class="n">marker_gene_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;gene&#39;</span><span class="p">:</span><span class="n">marker_gene</span><span class="p">,</span> <span class="s1">&#39;label&#39;</span><span class="p">:</span><span class="n">marker_gene_label</span><span class="p">})</span>
<span class="n">marker_gene_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">results_folder</span><span class="o">+</span><span class="s1">&#39;marker_gene.csv&#39;</span><span class="p">)</span>
<span class="n">adata_sc</span> <span class="o">=</span> <span class="n">adata_sc</span><span class="p">[:,</span> <span class="n">marker_gene</span><span class="p">]</span>
<span class="n">adata_st</span> <span class="o">=</span> <span class="n">adata_st</span><span class="p">[:,</span> <span class="n">marker_gene</span><span class="p">]</span>
</pre></div>
</div>
</div>
<p>If saving of the selected marker genes to file is not necessary, the code can be simplified to</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">marker_gene_dict</span> <span class="o">=</span> <span class="n">spotiphy</span><span class="o">.</span><span class="n">sc_reference</span><span class="o">.</span><span class="n">marker_selection</span><span class="p">(</span><span class="n">adata_sc</span><span class="p">,</span> <span class="n">key_type</span><span class="o">=</span><span class="n">key_type</span><span class="p">,</span> <span class="n">n_select</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">threshold_p</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
                                                          <span class="n">threshold_fold</span><span class="o">=</span><span class="mf">1.5</span><span class="p">,</span> <span class="n">q</span><span class="o">=</span><span class="mf">0.15</span><span class="p">)</span>
<span class="n">adata_sc</span> <span class="o">=</span> <span class="n">adata_sc</span><span class="p">[:,</span> <span class="n">marker_gene</span><span class="p">]</span>
<span class="n">adata_st</span> <span class="o">=</span> <span class="n">adata_st</span><span class="p">[:,</span> <span class="n">marker_gene</span><span class="p">]</span>
</pre></div>
</div>
<p>Construct the single_cell reference matrix based on the selected marker genes and visualize the results. We can examine which cell types are similar according to the heatmap.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">sc_ref</span> <span class="o">=</span> <span class="n">spotiphy</span><span class="o">.</span><span class="n">construct_sc_ref</span><span class="p">(</span><span class="n">adata_sc</span><span class="p">,</span> <span class="n">key_type</span><span class="o">=</span><span class="n">key_type</span><span class="p">)</span>
<span class="n">spotiphy</span><span class="o">.</span><span class="n">sc_reference</span><span class="o">.</span><span class="n">plot_heatmap</span><span class="p">(</span><span class="n">adata_sc</span><span class="p">,</span> <span class="n">key_type</span><span class="p">,</span> <span class="n">save</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">out_dir</span><span class="o">=</span><span class="n">results_folder</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
14it [00:00, 424.32it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/Spotiphy_tutorial_1_15_1.png" src="_images/Spotiphy_tutorial_1_15_1.png" />
</div>
</div>
</section>
<section id="Deconvolution:-Estimate-cell-abundance-in-spatial-spots">
<h2>Deconvolution: Estimate cell abundance in spatial spots<a class="headerlink" href="#Deconvolution:-Estimate-cell-abundance-in-spatial-spots" title="Link to this heading">ÔÉÅ</a></h2>
<p>For performing the deconvolution, we calculate the posterior distribution of certain parameters within a probabilistic generative model using <code class="docutils literal notranslate"><span class="pre">`Pyro</span></code> &lt;<a class="reference external" href="https://pyro.ai/">https://pyro.ai/</a>&gt;`__.</p>
<p>Parameters of the function <code class="docutils literal notranslate"><span class="pre">estimation_proportion</span></code>: - <code class="docutils literal notranslate"><span class="pre">X</span></code>: Spatial transcriptomics data. n_spot<em>n_gene. - ``adata_sc``: scRNA data (Anndata). - ``sc_ref``: Single cell reference. n_type</em>n_gene. - <code class="docutils literal notranslate"><span class="pre">type_list</span></code>: List of the cell types. - <code class="docutils literal notranslate"><span class="pre">key_type</span></code>: Column name of the cell types in adata_sc. - <code class="docutils literal notranslate"><span class="pre">batch_prior</span></code>: Parameter involved in the prior distribution of the batch effect factors. It is recommended to select from <span class="math notranslate nohighlight">\([1, 2]\)</span>. - <code class="docutils literal notranslate"><span class="pre">n_epoch</span></code>: Number of training epoch.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">device</span> <span class="o">=</span> <span class="s1">&#39;cuda&#39;</span> <span class="k">if</span> <span class="n">torch</span><span class="o">.</span><span class="n">cuda</span><span class="o">.</span><span class="n">is_available</span><span class="p">()</span> <span class="k">else</span> <span class="s1">&#39;cpu&#39;</span>
<span class="n">X</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">adata_st</span><span class="o">.</span><span class="n">X</span><span class="p">)</span>
<span class="n">cell_proportion</span> <span class="o">=</span> <span class="n">spotiphy</span><span class="o">.</span><span class="n">deconvolution</span><span class="o">.</span><span class="n">estimation_proportion</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="n">adata_sc</span><span class="p">,</span> <span class="n">sc_ref</span><span class="p">,</span> <span class="n">type_list</span><span class="p">,</span> <span class="n">key_type</span><span class="p">,</span> <span class="n">n_epoch</span><span class="o">=</span><span class="mi">8000</span><span class="p">,</span>
                                                               <span class="n">plot</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">batch_prior</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">device</span><span class="o">=</span><span class="n">device</span><span class="p">)</span>
<span class="n">adata_st</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">type_list</span><span class="p">]</span> <span class="o">=</span> <span class="n">cell_proportion</span>
<span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">results_folder</span><span class="o">+</span><span class="s1">&#39;proportion.npy&#39;</span><span class="p">,</span> <span class="n">cell_proportion</span><span class="p">)</span>  <span class="c1"># Save the proportion for future usage</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà| 8000/8000 [08:26&lt;00:00, 15.81it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/Spotiphy_tutorial_1_17_1.png" src="_images/Spotiphy_tutorial_1_17_1.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
2272758115.py (5): Trying to modify attribute `.obs` of view, initializing view as actual.
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">vmax</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">quantile</span><span class="p">(</span><span class="n">adata_st</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">type_list</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="mf">0.98</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">vmax</span><span class="p">[</span><span class="n">vmax</span><span class="o">&lt;</span><span class="mf">0.05</span><span class="p">]</span> <span class="o">=</span> <span class="mf">0.05</span>
<span class="k">with</span> <span class="n">mpl</span><span class="o">.</span><span class="n">rc_context</span><span class="p">({</span><span class="s1">&#39;figure.figsize&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">5</span><span class="p">],</span> <span class="s1">&#39;figure.dpi&#39;</span><span class="p">:</span> <span class="mi">400</span><span class="p">,</span> <span class="s1">&#39;xtick.labelsize&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">}):</span>
    <span class="n">ax</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spatial</span><span class="p">(</span><span class="n">adata_st</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;viridis&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">type_list</span><span class="p">,</span> <span class="n">img_key</span><span class="o">=</span><span class="s1">&#39;hires&#39;</span><span class="p">,</span> <span class="n">vmin</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">vmax</span><span class="p">),</span>
                       <span class="n">size</span><span class="o">=</span><span class="mf">1.3</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">alpha_img</span><span class="o">=</span><span class="mf">0.4</span><span class="p">)</span>
    <span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">get_figure</span><span class="p">()</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">results_folder</span><span class="o">+</span><span class="s1">&#39;Spotiphy.jpg&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/Spotiphy_tutorial_1_18_0.png" src="_images/Spotiphy_tutorial_1_18_0.png" />
</div>
</div>
</section>
<section id="Segmentation:-Estimate-cell-number-in-each-spatial-spot">
<h2>Segmentation: Estimate cell number in each spatial spot<a class="headerlink" href="#Segmentation:-Estimate-cell-number-in-each-spatial-spot" title="Link to this heading">ÔÉÅ</a></h2>
<p>We use the package <code class="docutils literal notranslate"><span class="pre">Stardist</span></code> to segmente nuclei from H&amp;E staining image. Then the number of nuclei in each spot is counted. Note that to get the results above, the image is not required.</p>
<p>Parameters use in function <code class="docutils literal notranslate"><span class="pre">Segmentation</span></code>:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">img</span></code> (numpy.ndarray):. Three channel stained image. In default, it should be hematoxylin and eosin (H&amp;E) stained image.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">spot_center</span></code>: Coordinates of the center of the spots.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">out_dir</span></code>: Output directory.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Prob_thresh</span></code>, <code class="docutils literal notranslate"><span class="pre">nms_thresh</span></code>: Two thresholds used in Stardist. User should adjust these two thresholds based on the segmentation results. More details can be found <a class="reference external" href="https://squidpy.readthedocs.io/en/stable/notebooks/tutorials/tutorial_stardist.html">here</a>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">spot_radius</span></code>: Radius of the spots. In 10X Visium, it should be 36.5.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">n_tiles</span></code>: Out of memory (OOM) errors can occur if the input image is too large. To avoid this problem, the input image is broken up into (overlapping) tiles that are processed independently and re-assembled. In default, we break the image into 8*8 tiles.</p></li>
</ul>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">results_folder</span><span class="o">+</span><span class="s1">&#39;segmentation/&#39;</span><span class="p">):</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">results_folder</span><span class="o">+</span><span class="s1">&#39;segmentation/&#39;</span><span class="p">)</span>
<span class="n">Segmentation</span> <span class="o">=</span> <span class="n">spotiphy</span><span class="o">.</span><span class="n">segmentation</span><span class="o">.</span><span class="n">Segmentation</span><span class="p">(</span><span class="n">img</span><span class="p">[:,</span> <span class="p">:,</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">]],</span> <span class="n">adata_st</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;spatial&#39;</span><span class="p">],</span>
                                                  <span class="n">n_tiles</span><span class="o">=</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span>
                                                  <span class="n">out_dir</span><span class="o">=</span><span class="n">results_folder</span><span class="o">+</span><span class="s1">&#39;segmentation/&#39;</span><span class="p">)</span>
<span class="n">Segmentation</span><span class="o">.</span><span class="n">segment_nucleus</span><span class="p">(</span><span class="n">save</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">n_cell_df</span> <span class="o">=</span> <span class="n">Segmentation</span><span class="o">.</span><span class="n">n_cell_df</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Suppress the output of tensorflow prediction for tensorflow version 2.12.0&gt;=2.9.0.
Found model &#39;2D_versatile_he&#39; for &#39;StarDist2D&#39;.
Loading network weights from &#39;weights_best.h5&#39;.
Loading thresholds from &#39;thresholds.json&#39;.
Using default values: prob_thresh=0.692478, nms_thresh=0.3.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà| 4/4 [00:07&lt;00:00,  1.78s/it]
</pre></div></div>
</div>
<p>Plot a small section of the segmented results for examination.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">Segmentation</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">save</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">crop</span><span class="o">=</span><span class="p">(</span><span class="mi">2000</span><span class="p">,</span> <span class="mi">2500</span><span class="p">,</span> <span class="mi">1000</span><span class="p">,</span> <span class="mi">1500</span><span class="p">),</span>
                  <span class="n">path</span><span class="o">=</span><span class="n">results_folder</span><span class="o">+</span><span class="s1">&#39;segmentation/segmentation_sample.jpg&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/Spotiphy_tutorial_1_22_0.png" src="_images/Spotiphy_tutorial_1_22_0.png" />
</div>
</div>
</section>
<section id="Decomposition:-Decompose-the-expression-of-each-spatial-spot-into-single-cell-level">
<h2>Decomposition: Decompose the expression of each spatial spot into single-cell level<a class="headerlink" href="#Decomposition:-Decompose-the-expression-of-each-spatial-spot-into-single-cell-level" title="Link to this heading">ÔÉÅ</a></h2>
<p>Decompose the spatial transcriptomics. Note that since we want to decompose all genes in the spatial transcriptomics, we reload the scRNA-seq data and spatial transcriptomics data. The argument <code class="docutils literal notranslate"><span class="pre">n_cell</span></code> represents the number of cells in each spot. When the histology image is not available, user can either estimate the cell number or simply set <code class="docutils literal notranslate"><span class="pre">n_cell</span></code> to <code class="docutils literal notranslate"><span class="pre">None</span></code>. When <code class="docutils literal notranslate"><span class="pre">n_cell</span></code> is <code class="docutils literal notranslate"><span class="pre">None</span></code>, we will assume that a cell type does not exist in a spot unless its proportion is larger than
<code class="docutils literal notranslate"><span class="pre">threshold</span></code>. For more details, please refer to the <a class="reference external" href="https://jyyulab.github.io/Spotiphy/deconvolution.html#deconvolution.decomposition">document</a>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">adata_sc_orig</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s2">&quot;data/scRNA_mouse_brain.h5ad&quot;</span><span class="p">)</span>
<span class="n">adata_st_orig</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s2">&quot;data/ST_mouse_brain.h5ad&quot;</span><span class="p">)</span>
<span class="n">adata_st_decomposed</span> <span class="o">=</span> <span class="n">spotiphy</span><span class="o">.</span><span class="n">deconvolution</span><span class="o">.</span><span class="n">decomposition</span><span class="p">(</span><span class="n">adata_st_orig</span><span class="p">,</span> <span class="n">adata_sc_orig</span><span class="p">,</span> <span class="n">key_type</span><span class="p">,</span>
                                                           <span class="n">cell_proportion</span><span class="p">,</span> <span class="n">save</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">out_dir</span><span class="o">=</span><span class="n">results_folder</span><span class="p">,</span>
                                                           <span class="n">verbose</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">spot_location</span><span class="o">=</span><span class="n">adata_st_orig</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;spatial&#39;</span><span class="p">],</span>
                                                           <span class="n">filtering_gene</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                                                           <span class="n">n_cell</span><span class="o">=</span><span class="n">n_cell_df</span><span class="p">[</span><span class="s1">&#39;cell_count&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
                                                           <span class="n">filename</span><span class="o">=</span><span class="s1">&#39;ST_decomposition.h5ad&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Prepared proportion data. Time use 0.05
Initialized scRNA and ST data. Time use 8.13
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
14it [00:00, 451.60it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Processed scRNA and ST data. Time use 0.22
Decomposition complete. Time use 0.07
Constructed ST decomposition data file. Time use 0.05
Saved file to output folder. Time use 0.35
</pre></div></div>
</div>
</section>
<section id="Generate-pseudo-single-cell-resolution-image">
<h2>Generate pseudo single-cell resolution image<a class="headerlink" href="#Generate-pseudo-single-cell-resolution-image" title="Link to this heading">ÔÉÅ</a></h2>
<p>As only the nuclei are stained in the H&amp;E staining image, we first infer the boundary of each cell.</p>
<p>Specifically, we gradually expend the area of each cell from the center of its nucleus. The expansion is stopped when: 1. the area of the cell is already larger than <code class="docutils literal notranslate"><span class="pre">max_area</span></code>, 2. the neighborhood are all occupied by other cells, or 3. the largest distance from the cell boundary to the nucleus center is larger than <code class="docutils literal notranslate"><span class="pre">max_dist</span></code>. <code class="docutils literal notranslate"><span class="pre">delta</span></code> represent the increase of expansion distance in each iteration.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">search_direction</span> <span class="o">=</span> <span class="p">[[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="p">[</span><span class="o">-</span><span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="o">-</span><span class="mi">2</span><span class="p">],</span>
                    <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">]]</span>
<span class="n">boundary_dict</span> <span class="o">=</span> <span class="n">spotiphy</span><span class="o">.</span><span class="n">segmentation</span><span class="o">.</span><span class="n">cell_boundary</span><span class="p">(</span><span class="n">Segmentation</span><span class="o">.</span><span class="n">nucleus_df</span><span class="p">[[</span><span class="s1">&#39;x&#39;</span><span class="p">,</span> <span class="s1">&#39;y&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
                                                    <span class="n">img_size</span><span class="o">=</span><span class="n">img</span><span class="o">.</span><span class="n">shape</span><span class="p">[:</span><span class="mi">2</span><span class="p">],</span> <span class="n">max_dist</span><span class="o">=</span><span class="mi">25</span><span class="p">,</span> <span class="n">max_area</span><span class="o">=</span><span class="mi">580</span><span class="p">,</span>
                                                    <span class="n">verbose</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">search_direction</span><span class="o">=</span><span class="n">search_direction</span><span class="p">,</span> <span class="n">delta</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà| 12/12 [00:07&lt;00:00,  1.56it/s]
100%|‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà| 4100/4100 [00:07&lt;00:00, 549.41it/s]
</pre></div></div>
</div>
<p>We can save the dictionary of the cell boundaries for future usage.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">results_folder</span><span class="o">+</span><span class="sa">f</span><span class="s1">&#39;segmentation/boundary_dict.pkl&#39;</span><span class="p">,</span> <span class="s1">&#39;wb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
    <span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">boundary_dict</span><span class="p">,</span> <span class="n">file</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Assign cell types to nuclei in the spatial tissue using these three steps:</p>
<ul class="simple">
<li><p>Function <code class="docutils literal notranslate"><span class="pre">proportion_to_count</span></code>: compute the count of each cell type using the segmentation outcomes and the estimated proportions.</p></li>
<li><p>Function <code class="docutils literal notranslate"><span class="pre">assign_type_spot</span></code>: assign cell types to nuclei within the spots.</p></li>
<li><p>Function <code class="docutils literal notranslate"><span class="pre">cell_proportion_smooth</span></code>: allocate cell types to nuclei located outside the spots.</p></li>
</ul>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">cell_number</span> <span class="o">=</span> <span class="n">spotiphy</span><span class="o">.</span><span class="n">deconvolution</span><span class="o">.</span><span class="n">proportion_to_count</span><span class="p">(</span><span class="n">cell_proportion</span><span class="p">,</span> <span class="n">n_cell_df</span><span class="p">[</span><span class="s1">&#39;cell_count&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
                                                         <span class="n">multiple_spots</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">Segmentation</span><span class="o">.</span><span class="n">nucleus_df</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">spotiphy</span><span class="o">.</span><span class="n">deconvolution</span><span class="o">.</span><span class="n">assign_type_spot</span><span class="p">(</span><span class="n">Segmentation</span><span class="o">.</span><span class="n">nucleus_df</span><span class="p">,</span> <span class="n">Segmentation</span><span class="o">.</span><span class="n">n_cell_df</span><span class="p">,</span>
                                            <span class="n">cell_number</span><span class="p">,</span> <span class="n">type_list</span><span class="p">))</span>
<span class="n">Segmentation</span><span class="o">.</span><span class="n">nucleus_df</span><span class="p">,</span> <span class="n">cell_proportion_smooth</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">spotiphy</span><span class="o">.</span><span class="n">deconvolution</span><span class="o">.</span><span class="n">assign_type_out</span><span class="p">(</span><span class="n">Segmentation</span><span class="o">.</span><span class="n">nucleus_df</span><span class="p">,</span> <span class="n">cell_proportion</span><span class="p">,</span>
                                           <span class="n">Segmentation</span><span class="o">.</span><span class="n">spot_center</span><span class="p">,</span> <span class="n">type_list</span><span class="p">,</span> <span class="n">band_width</span><span class="o">=</span><span class="mi">150</span><span class="p">))</span>
</pre></div>
</div>
</div>
<p>Define the class <code class="docutils literal notranslate"><span class="pre">plot_visium</span></code> for generating the pseudo image. Then we plot the legend of the images.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">plot_visium</span> <span class="o">=</span> <span class="n">spotiphy</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">Plot_Visium</span><span class="p">(</span><span class="n">segmentation</span><span class="o">=</span><span class="n">Segmentation</span><span class="p">,</span> <span class="n">boundary_dict</span><span class="o">=</span><span class="n">boundary_dict</span><span class="p">,</span>
                                        <span class="n">type_list</span><span class="o">=</span><span class="n">type_list</span><span class="p">)</span>
<span class="n">plot_visium</span><span class="o">.</span><span class="n">plot_legend</span><span class="p">(</span><span class="n">save</span><span class="o">=</span><span class="n">results_folder</span><span class="o">+</span><span class="s1">&#39;legend.png&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/Spotiphy_tutorial_1_33_0.png" src="_images/Spotiphy_tutorial_1_33_0.png" />
</div>
</div>
<p>We now generate the pseudo image. The user need to specify the following arguments: - <code class="docutils literal notranslate"><span class="pre">shape</span></code>: When adding the cells, which shape should be used. Should be selected from <code class="docutils literal notranslate"><span class="pre">cell</span></code>, <code class="docutils literal notranslate"><span class="pre">nucleus</span></code>, or <code class="docutils literal notranslate"><span class="pre">circle</span></code> - <code class="docutils literal notranslate"><span class="pre">cell</span></code>: Which group of cells should be included in the image: <code class="docutils literal notranslate"><span class="pre">in</span></code>, <code class="docutils literal notranslate"><span class="pre">out</span></code>, or <code class="docutils literal notranslate"><span class="pre">both</span></code>. - <code class="docutils literal notranslate"><span class="pre">boundary</span></code>: Which group of boundaries should be included in the image: <code class="docutils literal notranslate"><span class="pre">in</span></code>, <code class="docutils literal notranslate"><span class="pre">out</span></code>, <code class="docutils literal notranslate"><span class="pre">both</span></code> or <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<p>For example, if we want to generate a pseudo image with cells inside the spots, the following code can be used.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">plot_visium</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">background</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">save</span><span class="o">=</span><span class="n">results_folder</span><span class="o">+</span><span class="s1">&#39;pseudo_image_in.png&#39;</span><span class="p">,</span> <span class="n">spot_color</span><span class="o">=</span><span class="p">(</span><span class="mi">200</span><span class="p">,</span> <span class="mi">200</span><span class="p">,</span> <span class="mi">200</span><span class="p">),</span>
                 <span class="n">shape</span><span class="o">=</span><span class="s1">&#39;cell&#39;</span><span class="p">,</span> <span class="n">cell</span><span class="o">=</span><span class="s1">&#39;in&#39;</span><span class="p">,</span> <span class="n">boundary</span><span class="o">=</span><span class="s1">&#39;in&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Adding cells.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà| 4100/4100 [00:07&lt;00:00, 575.83it/s]
14it [00:00, 27.47it/s]
10747it [00:00, 1791652.49it/s]
100%|‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà| 10747/10747 [00:00&lt;00:00, 307073.12it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Adding spots.
Saving the image.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/Spotiphy_tutorial_1_35_3.png" src="_images/Spotiphy_tutorial_1_35_3.png" />
</div>
</div>
<p>If we want to show all the cells using circles without the cell boundary, we can use the following code:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">plot_visium</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">background</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">save</span><span class="o">=</span><span class="n">results_folder</span><span class="o">+</span><span class="s1">&#39;pseudo_image_both_circle.png&#39;</span><span class="p">,</span>
                 <span class="n">spot_color</span><span class="o">=</span><span class="p">(</span><span class="mi">200</span><span class="p">,</span> <span class="mi">200</span><span class="p">,</span> <span class="mi">200</span><span class="p">),</span> <span class="n">shape</span><span class="o">=</span><span class="s1">&#39;circle&#39;</span><span class="p">,</span> <span class="n">cell</span><span class="o">=</span><span class="s1">&#39;both&#39;</span><span class="p">,</span> <span class="n">boundary</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Adding cells.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
10747it [00:00, 182128.94it/s]
100%|‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà| 10747/10747 [00:00&lt;00:00, 2149350.81it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Adding spots.
Saving the image.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>

</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/Spotiphy_tutorial_1_37_4.png" src="_images/Spotiphy_tutorial_1_37_4.png" />
</div>
</div>
<p>Note that we can also save the updated class <code class="docutils literal notranslate"><span class="pre">Segmentation</span></code> for future usage.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">results_folder</span><span class="o">+</span><span class="sa">f</span><span class="s1">&#39;segmentation/Segmentation.pkl&#39;</span><span class="p">,</span> <span class="s1">&#39;wb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
    <span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">Segmentation</span><span class="p">,</span> <span class="n">file</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="tutorials.html" class="btn btn-neutral float-left" title="Tutorials" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="questions.html" class="btn btn-neutral float-right" title="Frequently Asked Questions" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Yu Lab, St. Jude Children&#39;s Research Hospital.</p>
  </div>

   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>